framework/resource/manager_test.go

```diff
@@ -298,3 +307,102 @@ func TestManagerDestroyAll_noDestroyFunc(t *testing.T) {
 	// Ensure we destroyed
 	require.Equal([]string{"B"}, destroyOrder)
 }
+
+// TestStatus_Manager tests the Manager's ability to call resource status
+// methods and present them for creating a report
+func TestStatus_Manager(t *testing.T) {
+	require := require.New(t)
+
+	init := func() *Manager {
+		return NewManager(
+			// state with status
+			WithResource(NewResource(
+				WithName("A"),
+				WithState(&testState{}),
+				WithCreate(func(s *testState, v int) error {
+					s.Value = v
+					return nil
+				}),
+				WithStatus(func(s *testState, sr *StatusResponse) error {
+					rr := &pb.StatusReport_Resource{
+						Name: fmt.Sprintf(statusNameTpl, s.Value),
+					}
+					sr.Reports = append(sr.Reports, rr)
+					return nil
+				}),
+			)),
+
+			// no state, with status
+			WithResource(NewResource(
+				WithName("B"),
+				WithCreate(func(s *testState) error {
+					// no-op
+					return nil
+				}),
+				WithStatus(func(sr *StatusResponse) error {
+					rr := &pb.StatusReport_Resource{
+						Name: "no state here",
+					}
+					sr.Reports = append(sr.Reports, rr)
+					return nil
+				}),
+			)),
+			WithResource(NewResource(
+				WithName("C"),
+				WithState(&testState2{}),
+				WithCreate(func(s *testState2, vs string) error {
+					v, _ := strconv.Atoi(vs)
+					s.Value = v
+					return nil
+				}),
+				WithStatus(func(s *testState2, sr *StatusResponse) error {
+					rr := &pb.StatusReport_Resource{
+						Name: fmt.Sprintf(statusNameTpl, s.Value),
+					}
+					// make sure we can return more than 1 StatusReport_Resource
+					// in a single Resource Status method
+					rr2 := &pb.StatusReport_Resource{
+						Name: fmt.Sprintf(statusNameTpl, s.Value+1),
+					}
+					sr.Reports = append(sr.Reports, rr, rr2)
+					return nil
+				}),
+			)),
+			// state, no status
+			WithResource(NewResource(
+				WithName("D"),
+				WithState(&testState3{}),
+				WithCreate(func(s *testState3) error {
+					s.Value = 0
+					return nil
+				}),
+			)),
```

Comment:
I was curious why these tests weren't hitting the infinite loop that I found in practice. I'm still not sure why, but if you remove this resource (`D`) that doesn't have a status func, it reproduces the loop when running against b5c7228a8bfe3f5f10919cc5698d6c831a2cad1d (i.e. before you applied the fixes). Trimming this down to just `A`, `B`, or `C` also does it (cc @mitchellh in case you're looking for an easy repro for an argmapper unit test)

I think it would make sense to include one more simplified test that has only one resource (maybe that looks like `C`), because that's what most plugins will have in practice, and it'll catch this problem if it ever regresses.

Reply:
Sounds good; I added an additional test in https://github.com/hashicorp/waypoint-plugin-sdk/pull/43/commits/e4ce5d48a5707780f5254c1eea0ab15070d89542